### PR TITLE
vscode: add experimental esbuild support

### LINKED ISSUE #28823: VSCE: Migrate from Webpack to esbuild
Our VS Code extension's build times are continuously slowing down as we get closer to v1. We find ourselves having to build more and more bundles to support new features (e.g. main extension bundle for web/web worker VS Code _and_ desktop/node VS Code, multiple webview bundles that pull from massive amounts of shared code), which contributes to long build times. To improve developer experience, we should try to migrate to esbuild.

We already have experimental support for esbuild for [the web app](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/doc/dev/background-information/web/build.md?L54:5), with great improvement over Webpack (~3s vs ~53s). We should build off this experimental support as well as add support for production builds.

### PR DESCRIPTION
**Pull request by tjkandala**
Closes #28823
 
Adds ability to build the VS Code extension with esbuild. Production build time has gone from 100 seconds to 1.5 seconds on my M1 Pro MBP! Moves and modifies experimental web esbuild plugins to work with the VS Code extension.

| esbuild | webpack |
| --- | --- |
| Screen Shot 2022-04-18 at 10 55 36 AM | Screen Shot 2022-04-18 at 12 03 48 PM |

Screen Shot 2022-04-18 at 10 55 22 AM

TODO: 
- ensure no regressions with integration tests (create separate issue)

## Test plan

- Steps for node extension host
  - Run `yarn build:esbuild` in the `client/vscode` directory
  - Run `"Launch VS Code Extension"` task in VS Code
  - Note that search and remote file browsing still work
- Steps for webworker extension host
  - Run `yarn build:esbuild:web` in the `client/vscode` directory
  - Run `"Launch VS Code Web Extension"` task in VS Code
  - Note that search and remote file browsing still work

## App preview:

- [Link](https://sg-web-tjk-vsce-esbuild.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-evixyevufh.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

File: client/vscode/package.json [hunk 1]
```diff
@@ -219,6 +219,9 @@
     "test": "ts-node ./tests/runTests.ts",
     "package": "ts-node ./scripts/package.ts",
     "task:gulp": "cross-env NODE_OPTIONS=\"--max_old_space_size=8192\" gulp",
+    "build:esbuild": "NODE_ENV=development yarn task:gulp esbuild",
+    "build:esbuild:web": "NODE_ENV=development TARGET_TYPE=webworker yarn task:gulp esbuild",
+    "watch:esbuild": "NODE_ENV=development WATCH=true yarn task:guild esbuild",
```
Comment (abeatrix) [inline client/vscode/package.json:224]:
```suggestion
    "watch:esbuild": "NODE_ENV=development WATCH=true yarn task:gulp esbuild",
```

**Review by abeatrix — APPROVED** [review]
A small typo that results in running the `yarn watch:esbuild` command to fail with `Error: Command "task:guild" not found.` Other than that is a great change to replace the webpack build, it's SO MUCH FASTER!! Thank you TJ!

File: client/vscode/src/webview/search-panel/components/BrandHeader.tsx [hunk 2]
```diff
@@ -16,7 +16,7 @@ export const BrandHeader: React.FunctionComponent<Pick<WebviewPageProps, 'theme'
             alt="Sourcegraph logo"
         />
         <div data-testid="brand-header" className={classNames(styles.logoText)}>
-            Search your code and 2M+ open source repositories
+            Search your code and 2M+ open source repositories WITH ESBUILD!
```
Comment (abeatrix) [inline client/vscode/src/webview/search-panel/components/BrandHeader.tsx:19]:
```suggestion
            Search your code and 2M+ open source repositories
```

**Review by abeatrix — CHANGES_REQUESTED** [review]
Looks like the current build for VS Code Web does not work

image

**Comment by abeatrix** [comment]
Should we document this somewhere? 

```
Steps for node extension host
1. Run yarn build:esbuild in the client/vscode directory
2. Run "Launch VS Code Extension" task in VS Code

Steps for webworker extension host
1. Run yarn build:esbuild:web in the client/vscode directory
2. Run "Launch VS Code Web Extension" task in VS Code

```

**Comment by tjkandala** [comment]
@vdavid I've merged the esbuild PR!

**Pull request merged**